cli/cli.go

```diff
@@ -0,0 +1,176 @@
+package cli
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/exercism/cli/debug"
+	update "github.com/inconshreveable/go-update"
+)
+
+var (
+	// BuildOS is the operating system (GOOS) used during the build process.
+	BuildOS string
+	// BuildARM is the ARM version (GOARM) used during the build process.
+	BuildARM string
+	// BuildARCH is the architecture (GOARCH) used during the build process.
+	BuildARCH string
+)
+
+var (
+	osMap = map[string]string{
+		"darwin":  "mac",
+		"linux":   "linux",
+		"windows": "windows",
+	}
+
+	archMap = map[string]string{
+		"amd64": "64bit",
+		"386":   "32bit",
+		"arm":   "arm",
+	}
+)
+
+var (
+	HTTPClient       = &http.Client{Timeout: 10 * time.Second}
+	LatestReleaseURL = "https://api.github.com/repos/exercism/cli/releases/latest"
+)
+
+type CLI struct {
+	Version       string
+	LatestRelease *Release
+}
+
+func New(version string) *CLI {
+	return &CLI{
+		Version: version,
+	}
+}
+
+func (c *CLI) IsUpgradeNeeded() (bool, error) {
+	if c.LatestRelease == nil {
+		resp, err := HTTPClient.Get(LatestReleaseURL)
+		if err != nil {
+			return false, err
+		}
+		defer resp.Body.Close()
+
+		var rel Release
+		if err := json.NewDecoder(resp.Body).Decode(&rel); err != nil {
+			return false, err
+		}
+		c.LatestRelease = &rel
+	}
+
+	rv, err := semver.Make(c.LatestRelease.Version())
+	if err != nil {
+		return false, fmt.Errorf("unable to parse latest version (%s): %s", c.LatestRelease.Version(), err)
+	}
+	cv, err := semver.Make(c.Version)
+	if err != nil {
+		return false, fmt.Errorf("unable to parse current version (%s): %s", c.Version, err)
+	}
+
+	return rv.GT(cv), nil
+}
+
+// Upgrade allows the user to upgrade to the latest version of the CLI.
+func (c *CLI) Upgrade() error {
```

Comment:
I moved the `Upgrade()` function directly onto the `CLI` type. The logic is the same as it was, but I moved some of the `fmt.Println`s to keep them in the `cmd` package.